File: next-permutation/next-permutation.java
public class Solution {
    public void nextPermutation(int[] nums) {
        int p1 = nums.length - 2;
        
        while (p1 >= 0 && nums[p1+1] <= nums[p1]) {
            p1--;
        }
        
        if (p1 >= 0) {
            int p2 = nums.length - 1;
            while (nums[p2] <= nums[p1]) {
                p2--;
            }
            swap(nums, p1, p2);
        }
        reverse(nums, p1 + 1);
    }
    
    private void reverse(int[] nums, int left) {
        int right = nums.length - 1;
        
        while (left < right) {
            swap(nums, left++, right--);
        }
    }
    
    private void swap(int[] nums, int left, int right) {
        int temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
    }
}

File: next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int start = nums.size() - 1;
        int end = nums.size() - 1;
    
        while (start > 0) {
            if (nums[start] > nums[start - 1]) {
                reverse(nums.begin() + start, nums.end());
                int pivot = start;
                while (pivot <= end && nums[pivot] <= nums[start - 1]) {
                    pivot++;
                }
                swap(nums[start - 1], nums[pivot]);
                return;
            }
            else {
                start--;
            }
        }
    
        reverse(nums.begin(), nums.end());
    }
};